pkg/media/whip/whipsrc.go

```diff
@@ -0,0 +1,371 @@
+	gatherComplete := webrtc.GatheringCompletePromise(s.pc)
+
+	// Sets the LocalDescription, and starts our UDP listeners
+	if err = s.pc.SetLocalDescription(answer); err != nil {
+		s.pc.Close()
+		return "", err
+	}
+
+	select {
+	case <-gatherComplete:
+		// success
+	case <-ctx.Done():
+		return "", psrpc.NewErrorf(psrpc.DeadlineExceeded, "timed out while waiting for ICE candidate gathering")
+	}
+
+	sdpAnswer := s.pc.LocalDescription().SDP
+
+	return sdpAnswer, nil
+}
+
+func (s *WHIPSource) postSDPAnswer(ctx context.Context, sdpAnswer string) error {
+	c := &http.Client{}
+	req, err := http.NewRequestWithContext(ctx, "POST", s.params.RelayUrl, bytes.NewReader([]byte(sdpAnswer)))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/sdp")
+
+	resp, err := c.Do(req)
+	switch {
+	case err != nil:
+		return psrpc.NewErrorf(psrpc.Internal, "failed connecting to relay")
+	case resp.StatusCode != http.StatusOK:
+		return psrpc.NewErrorf(psrpc.Internal, "failed response from relay")
+	}
+
+	return nil
+}
+
+func (s *WHIPSource) addTrack(track *webrtc.TrackRemote, receiver *webrtc.RTPReceiver) {
+	kind := streamKindFromCodecType(track.Kind())
+	logger.Infow("track has started", "type", track.PayloadType(), "codec", track.Codec().MimeType, "kind", kind)
+
+	s.trackLock.Lock()
+	defer s.trackLock.Unlock()
+	s.tracks[track.ID()] = track
+	if _, ok := s.trackSrc[kind]; ok {
+		logger.Warnw("duplicate track of the same kind", errors.ErrUnsupportedDecodeFormat, "kind", kind)
+		return
+	}
+
+	sync := s.sync.AddTrack(track, whipIdentity)
+	appSrc, err := NewWHIPAppSource(track, receiver, sync, s.writePLI, s.sync.OnRTCP)
+	if err != nil {
+		logger.Warnw("failed creating whip app source", err)
+		return
+	}
+	s.trackSrc[kind] = appSrc
+
+	select {
+	case s.trackAddedChan <- track:
+	default:
+		logger.Warnw("failed notifying of new track", errors.New("channel full"))
+	}
+}
+
+func streamKindFromCodecType(typ webrtc.RTPCodecType) types.StreamKind {
+	switch typ {
+	case webrtc.RTPCodecTypeAudio:
+		return types.Audio
+	case webrtc.RTPCodecTypeVideo:
+		return types.Video
+	default:
+		return types.Unknown
+	}
+}
+
+func (w *WHIPSource) writePLI(ssrc webrtc.SSRC) {
+	logger.Debugw("sending PLI request", "ssrc", ssrc)
+	pli := []rtcp.Packet{
+		&rtcp.PictureLossIndication{SenderSSRC: uint32(ssrc), MediaSSRC: uint32(ssrc)},
+	}
+	err := w.pc.WriteRTCP(pli)
+	if err != nil {
+		logger.Warnw("failed writing PLI", err, "ssrc", ssrc)
+	}
+}
+
+func getExpectedTrackCount(offer *webrtc.SessionDescription) (int, error) {
+	parsed, err := offer.Unmarshal()
+	if err != nil {
+		return 0, err
+	}
+
+	return len(parsed.MediaDescriptions), nil
+}
+
+func newMediaEngine(p *params.Params) (*webrtc.MediaEngine, error) {
+	m := &webrtc.MediaEngine{}
+
+	for _, codec := range []webrtc.RTPCodecParameters{
+		{
+			RTPCodecCapability: webrtc.RTPCodecCapability{MimeType: webrtc.MimeTypePCMA, ClockRate: 8000},
```

Comment:
I didn't realize WHIP supported PCM.

Reply:
The spec says nothing about codec support (which is one more reason that makes not transcoding painful operationally). I do not know if any whip client currently supports PCM. Dean had PCM in his prototype, so I put there too. We can remove it if we want to be more conservative about codec support initially.